### PR TITLE
Add support for applying setting filters when displaying repository settings (Backport to 1.6)

### LINKED ISSUE #11265: Snapshot/Restore: Add support for applying setting filters when displaying repository settings
Currently all settings that were specified during repository creation are displayed. We should allow plugins such as cloud-aws to filter out sensitive settings from the repository. See elastic/elasticsearch-cloud-aws/issues/184 for an example.

### PR DESCRIPTION
**Pull request by imotov**
Currently all settings that were specified during repository creation are displayed. This commit enables  plugins such as cloud-aws to filter out sensitive settings from the repository.

Closes #11265

### CODE REVIEW

File: src/main/java/org/elasticsearch/cluster/ClusterState.java [hunk 1]
```diff
@@ -451,7 +451,12 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
 
             for (ObjectObjectCursor<String, MetaData.Custom> cursor : metaData.customs()) {
                 builder.startObject(cursor.key);
-                MetaData.lookupFactorySafe(cursor.key).toXContent(cursor.value, builder, params);
+                MetaData.Custom.Factory<MetaData.Custom> factory = MetaData.lookupFactorySafe(cursor.key);
+                if (factory instanceof MetaData.Custom.ToXFilteredContent) {
```
Comment (dakrone) [inline src/main/java/org/elasticsearch/cluster/ClusterState.java]: Maybe add a comment here about why this is/was needed?

File: src/main/java/org/elasticsearch/cluster/metadata/MetaData.java [hunk 2]
```diff
@@ -93,6 +94,10 @@
                 return API_ONLY;
             }
         }
+
+        interface ToXFilteredContent<T extends Custom> {
```
Comment (dakrone) [inline src/main/java/org/elasticsearch/cluster/metadata/MetaData.java]: Needs a javadoc to explain why it's different (the settingsfilter)

**Comment by dakrone** [comment]
LGTM, left some super minor comments about docs

**Comment by imotov** [comment]
Merged b76c1595f41b5144729d32dcc2dee1402940fce9

**Pull request closed without merging**